services/worker/src/worker/job_runners/split/descriptive_stats.py

```diff
@@ -0,0 +1,352 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2023 The HuggingFace Authors.
+
+import logging
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple, TypedDict, Union
+
+import duckdb
+import numpy as np
+from libcommon.constants import PROCESSING_STEP_SPLIT_DESCRIPTIVE_STATS_VERSION
+from libcommon.exceptions import (
+    NoSupportedFeaturesError,
+    ParquetResponseEmptyError,
+    PreviousStepFormatError,
+    SplitWithTooBigParquetError,
+    StatsComputationError,
+)
+from libcommon.processing_graph import ProcessingStep
+from libcommon.simple_cache import get_previous_step_or_raise
+from libcommon.storage import StrPath
+from libcommon.utils import JobInfo
+from tqdm import tqdm
+
+from worker.config import AppConfig
+from worker.dtos import CompleteJobResult
+from worker.job_runners.split.split_job_runner import SplitJobRunnerWithCache
+from worker.utils import check_split_exists
+
+PARQUET_FILENAME = "dataset.parquet"
+
+DECIMALS = 5
+
+INTEGER_DTYPES = ["int8", "int16", "int32", "int64"]
+FLOAT_DTYPES = ["float16", "float32", "float64"]
+NUMERICAL_DTYPES = INTEGER_DTYPES + FLOAT_DTYPES
+
+
+class Histogram(TypedDict):
+    hist: List[int]
+    bin_edges: List[float]
+
+
+class NumericalStatsItem(TypedDict):
+    nan_count: int
+    nan_prop: float
+    min: float
+    max: float
+    mean: float
+    median: float
+    std: float
+    histogram: Histogram
+
+
+class CategoricalStatsItem(TypedDict):
+    nan_count: int
+    nan_prop: float
+    n_unique: int
+    frequencies: Dict[str, int]
+
+
+class StatsPerColumnItem(TypedDict):
+    column_name: str
+    column_type: str
+    column_dtype: Optional[str]
+    column_stats: Union[NumericalStatsItem, CategoricalStatsItem]
+
+
+class SplitDescriptiveStatsResponse(TypedDict):
+    num_examples: int
+    stats: List[StatsPerColumnItem]
+
+
+def compute_histogram(
+    con: duckdb.DuckDBPyConnection,
+    column_name: str,
+    parquet_filename: Path,
+    bin_size: int,
+    min_value: Union[int, float],
+    n_bins: int,
+    n_samples: Optional[int] = None,
+) -> Histogram:
+    hist_query = f"""
```

Comment:
also can't figure out how to avoid possible sql injection here and in all queries below. I need to pass `column_name` as a parameter, but when doing something like 
```
con.execute("select $column_name" from df ...", {"column_name": column_name})
```
`column_name` works like a string value, not like a column reference (like it's `select 'column_name' from df`, not `select column_name from df`).

Reply:
> I think it's not really possible to pass anything bad in column names of a dataset or a parquet filename

no, actually it is, I thought we don't allow spaces and other symbols in column names but apparently we do :D 

![image](https://github.com/huggingface/datasets-server/assets/16348744/495fad21-c833-4d2d-b12a-c0862cc9b458)

Reply:
> no, actually it is, I thought we don't allow spaces and other symbols in column names but apparently we do :D

Ah nice xD But yea anyway we can see later how to fix the sql, it's not that important imo (just add a TODO maybe ?)